src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs

```diff
@@ -204,5 +204,65 @@ public void PaxNameCollision_DedupInExtendedAttributes()
             Assert.True(File.Exists(path1));
             Assert.True(Path.Exists(path2));
         }
+
+        [Theory]
+        [MemberData(nameof(GetTestTarFormats))]
+        public void UnseekableStreams_RoundTrip(TestTarFormat testFormat)
+        {
+            using TempDirectory root = new();
+
+            using MemoryStream sourceStream = GetTarMemoryStream(CompressionMethod.Uncompressed, testFormat, "many_small_files");
+            using WrappedStream sourceUnseekableArchiveStream = new(sourceStream, canRead: true, canWrite: false, canSeek: false);
+
+            TarFile.ExtractToDirectory(sourceUnseekableArchiveStream, root.Path, overwriteFiles: false);
+
+            using MemoryStream destinationStream = new();
+            using WrappedStream destinationUnseekableArchiveStream = new(destinationStream, canRead: true, canWrite: true, canSeek: false);
+            TarFile.CreateFromDirectory(root.Path, destinationUnseekableArchiveStream, includeBaseDirectory: false);
+
+            FileSystemEnumerable<FileSystemInfo> fileSystemEntries = new FileSystemEnumerable<FileSystemInfo>(
+                directory: root.Path,
+                transform: (ref FileSystemEntry entry) => entry.ToFileSystemInfo(),
+                options: new EnumerationOptions() { RecurseSubdirectories = true });
+
+            destinationStream.Position = 0;
+            using TarReader reader = new TarReader(destinationStream, leaveOpen: false);
+
+            // Size of files in many_small_files.tar are expected to be tiny and all equal
+            int bufferLength = 1024;
+            byte[] buffer1 = new byte[bufferLength];
+            byte[] buffer2 = new byte[bufferLength];
```

Comment:
these names don't tell me much. They could be renamed to express their job

```suggestion
            byte[] fileContent = new byte[bufferLength];
            byte[] dataStreamContent = new byte[bufferLength];
```